File: apps/planner-api/src/security/unicode.py
"""Unicode scalar sanitization for planner-gateway transport.

OWNS: Replacing unpaired UTF-16 surrogates with U+FFFD before UTF-8 encode.
WHY: JSON may legally carry lone surrogates from page-derived JS strings. Python's
     UTF-8 encoder rejects them (`surrogates not allowed`), which previously
     crashed Gemini request serialization.
MUST: Preserve valid Unicode (emoji, Devanagari, CJK, combining marks).
MUST NOT: Strip emoji, force ASCII, log the offending string, or broaden egress.
"""

from __future__ import annotations

import json
from typing import Any

UNICODE_REPLACEMENT = "\ufffd"


def sanitize_unicode(text: str) -> str:
    """Replace code points in U+D800..U+DFFF with U+FFFD. Valid scalars unchanged."""
    if not text:
        return text
    return "".join(
        ch if _is_unicode_scalar(ord(ch)) else UNICODE_REPLACEMENT for ch in text
    )


def _is_unicode_scalar(code: int) -> bool:
    return 0 <= code <= 0xD7FF or 0xE000 <= code <= 0x10FFFF


def sanitize_json_value(value: Any) -> Any:
    """Recursively sanitize strings in JSON-like structures."""
    if isinstance(value, str):
        return sanitize_unicode(value)
    if isinstance(value, list):
        return [sanitize_json_value(item) for item in value]
    if isinstance(value, dict):
        return {
            sanitize_unicode(str(key)): sanitize_json_value(child)
            for key, child in value.items()
        }
    return value


def encode_json_utf8(value: Any) -> bytes:
    """Serialize to UTF-8 bytes. Callers must sanitize first; this still uses strict UTF-8."""
    return json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")

File: packages/protocol/src/unicode.ts
/**
 * Unicode scalar sanitization (transport safety).
 *
 * OWNS: Deterministic replacement of unpaired UTF-16 surrogates with U+FFFD.
 * WHY: JSON can legally carry `\ud83d`. Python/UTF-8 cannot encode that code unit,
 *      which previously crashed Gemini serialization on real pages (YouTube).
 * MUST: Preserve every valid Unicode scalar (ASCII, Latin, emoji, Devanagari, CJK,
 *       combining marks, supplementary-plane characters).
 * MUST NOT: Strip emoji, force ASCII, or drop the surrounding SafeContext.
 */

export const UNICODE_REPLACEMENT = '\uFFFD';

function isHighSurrogate(code: number): boolean {
  return code >= 0xd800 && code <= 0xdbff;
}

function isLowSurrogate(code: number): boolean {
  return code >= 0xdc00 && code <= 0xdfff;
}

/**
 * Replace unpaired UTF-16 surrogates with U+FFFD. Valid scalar sequences are unchanged.
 * Deterministic: the same input always yields the same output.
 */
export function sanitizeUnicodeScalars(input: string): string {
  if (!input) return input;
  let out = '';
  for (let i = 0; i < input.length; i += 1) {
    const code = input.charCodeAt(i);
    if (isHighSurrogate(code)) {
      const next = i + 1 < input.length ? input.charCodeAt(i + 1) : 0;
      if (isLowSurrogate(next)) {
        out += input[i];
        out += input[i + 1];
        i += 1;
        continue;
      }
      out += UNICODE_REPLACEMENT;
      continue;
    }
    if (isLowSurrogate(code)) {
      out += UNICODE_REPLACEMENT;
      continue;
    }
    out += input[i];
  }
  return out;
}

/**
 * Recursively sanitize strings in JSON-like values. Keys are sanitized too.
 * Non-string primitives and null are preserved.
 */
export function sanitizeUnicodeDeep<T>(value: T): T {
  if (typeof value === 'string') {
    return sanitizeUnicodeScalars(value) as T;
  }
  if (Array.isArray(value)) {
    return value.map((item) => sanitizeUnicodeDeep(item)) as T;
  }
  if (value && typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const out: Record<string, unknown> = {};
    for (const [key, child] of Object.entries(record)) {
      out[sanitizeUnicodeScalars(key)] = sanitizeUnicodeDeep(child);
    }
    return out as T;
  }
  return value;
}
